mempool/mempool.go

```diff
@@ -0,0 +1,253 @@
+				return err
+			}
+		} else {
+			// empty list, make new item both the head and the tail
+			if err = p.updateHead(txn, userTxn.Transaction.Hash()); err != nil {
+				return err
+			}
+		}
+
+		if err = p.updateTail(txn, userTxn.Transaction.Hash()); err != nil {
+			return err
+		}
+
+		pLen, err := p.len(txn)
+		if err != nil {
+			return err
+		}
+		return p.updateLen(txn, pLen+1) // don't worry about overflows, highly unlikely
+	}); err != nil {
+		return err
+	}
+
+	select {
+	case p.txPushed <- struct{}{}:
+	default:
+	}
+
+	return nil
+}
+
+// Pop returns the transaction with the highest priority from the pool
+func (p *Pool) Pop() (BroadcastedTransaction, error) {
+	var nextTxn BroadcastedTransaction
+	return nextTxn, p.db.Update(func(txn db.Transaction) error {
+		headHash, err := p.headHash(txn)
+		if err != nil {
+			return err
+		}
+
+		headElem, err := p.elem(txn, headHash)
+		if err != nil {
+			return err
+		}
+
+		if err = txn.Delete(headHash.Marshal()); err != nil {
+			return err
+		}
+
+		if headElem.NextHash == nil {
+			// the list is empty now
+			if err = txn.Delete([]byte(headKey)); err != nil {
+				return err
+			}
+			if err = txn.Delete([]byte(tailKey)); err != nil {
+				return err
+			}
+		} else {
+			if err = p.updateHead(txn, headElem.NextHash); err != nil {
+				return err
+			}
+		}
+
+		pLen, err := p.len(txn)
+		if err != nil {
+			return err
+		}
+
+		if err = p.updateLen(txn, pLen-1); err != nil {
+			return err
+		}
+		nextTxn = headElem.Txn
+		return nil
+	})
+}
+
+// Remove removes a set of transactions from the pool
+func (p *Pool) Remove(hash ...*felt.Felt) error {
+	return errors.New("not implemented")
+}
+
+// Len returns the number of transactions in the pool
+func (p *Pool) Len() (uint64, error) {
+	var l uint64
+	return l, p.db.View(func(txn db.Transaction) error {
+		var err error
+		l, err = p.len(txn)
+		return err
+	})
+}
+
+func (p *Pool) Wait() <-chan struct{} {
+	return p.txPushed
+}
+
+func (p *Pool) len(txn db.Transaction) (uint64, error) {
+	var l uint64
+	err := txn.Get([]byte(poolLengthKey), func(b []byte) error {
+		l = binary.BigEndian.Uint64(b)
+		return nil
+	})
+
+	if err != nil && errors.Is(err, db.ErrKeyNotFound) {
+		return 0, nil
+	}
+	return l, err
+}
+
+func (p *Pool) updateLen(txn db.Transaction, l uint64) error {
+	return txn.Set([]byte(poolLengthKey), binary.BigEndian.AppendUint64(nil, l))
+}
+
+func (p *Pool) headHash(txn db.Transaction) (*felt.Felt, error) {
+	var head *felt.Felt
+	return head, txn.Get([]byte(headKey), func(b []byte) error {
+		head = new(felt.Felt).SetBytes(b)
+		return nil
+	})
+}
+
+func (p *Pool) updateHead(txn db.Transaction, head *felt.Felt) error {
+	return txn.Set([]byte(headKey), head.Marshal())
+}
+
+func (p *Pool) tailHash(txn db.Transaction) (*felt.Felt, error) {
+	var tail *felt.Felt
+	return tail, txn.Get([]byte(tailKey), func(b []byte) error {
+		tail = new(felt.Felt).SetBytes(b)
+		return nil
+	})
+}
```

Comment:
Yup, implemented this change (and a similar one for headHash)

Reply:
Also, just fyi, `new(felt.Felt)` seems to cause a heap allocation

Reply:
The same could be applied to `headHash`. Think that this optimization is most important with transactions that are getting executed a lot. In the previous example, with `New` for `Pool` then if it only executed once it will make no real difference; on the other hand I believe this will be executed on a per transaction basis.